services/streammanager/googlecloudfunction/googlecloudfunction.go

```diff
@@ -0,0 +1,215 @@
+}
+
+func (producer *GoogleCloudFunctionProducer) Produce(jsonData json.RawMessage, destConfig Config) (statusCode int, respStatus, responseMessage string) {
+
+	parsedJSON := gjson.ParseBytes(jsonData)
+	if destConfig.FunctionEnvironment == "gen1" && destConfig.RequireAuthentication {
+		client := producer.client
+
+		if client == nil {
+			respStatus = "Failure"
+			responseMessage = "[GoogleCloudFunction] error  :: Failed to initialize GoogleCloudFunction client"
+			return 400, respStatus, responseMessage
+		}
+
+		functionName := getFunctionName(destConfig.GoogleCloudFunctionUrl)
+
+		fmt.Print(functionName)
+
+		requestPayload := &cloudfunctions.CallFunctionRequest{
+			Data: `{"input_key": "input_value"}`,
+		}
+
+		// Make the HTTP request
+		call := client.service.Projects.Locations.Functions.Call(functionName, requestPayload)
+
+		response, err := call.Do()
+		if err != nil {
+			if apiErr, ok := err.(*googleapi.Error); ok && apiErr.Code == http.StatusNotModified {
+				fmt.Println("Function call was not executed (Not Modified)")
+			} else {
+				log.Fatalf("Failed to call function: %v", err)
+			}
+		}
+
+		// Process the response (sample response handling here).
+		if response != nil {
+			fmt.Printf("Function call status code: %d\n", response.HTTPStatusCode)
+			// Handle response content as needed.
+		}
+		fmt.Println("Request successful!")
+
+		respStatus = "Success"
+		responseMessage = "[GoogleCloudFunction] :: Message Payload inserted with messageId :: "
+		return 200, respStatus, responseMessage
+	}
+
+	ctx := context.Background()
+	url := destConfig.GoogleCloudFunctionUrl // Replace with your actual audience URL
+
+	// Construct the GoogleCredentials object which obtains the default configuration from your
+	// working environment.
+	ts, err := idtoken.NewTokenSource(ctx, url, option.WithCredentialsJSON([]byte(destConfig.Credentials)))
+	if err != nil {
+		fmt.Print("failed to create NewTokenSource: %w", err)
+	}
+
+	// Get the ID token, to make an authenticated call to the target audience.
+	token, err := ts.Token()
+	if err != nil {
+		fmt.Print("failed to receive token: %w", err)
+	}
+
+	// Create a POST request
+	req, err := http.NewRequest("POST", destConfig.GoogleCloudFunctionUrl, bytes.NewBuffer(jsonData))
+	if err != nil {
+		fmt.Println("Error creating request:", err)
+		return
+	}
+
+	// Set the appropriate headers
+	req.Header.Set("Content-Type", "application/json")
+	if destConfig.RequireAuthentication {
+		req.Header.Add("Authorization", "Bearer "+token.AccessToken)
+	}
+
+	// Make the request using the client
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		fmt.Println("Error making request:", err)
+		return
+	}
+
+	if resp.Status == "OK" {
+		respStatus = "Success"
+		responseMessage = "[GoogleSheets] :: Message Payload inserted with messageId :: " + parsedJSON.Get("id").String()
+	}
+	fmt.Print(resp.Status)
+
+	return 200, respStatus, responseMessage
```

Comment:
Can we also put this in a separate function ?

Reply:
Done